src/VectorStore.ts

```diff
@@ -1,186 +1,292 @@
-		const results: Array<[number, string, Vector]> = []
-		
-		for (const entry of this.filePathToVector.entries()) {
-			const cosineSimilarity = similarity(searchVector, entry[1])
-			results.push([cosineSimilarity, entry[0], entry[1]])
+	private async generateFileEntry(fileContents: string, mtime: number, md5hash: string): Promise<FileEntry | undefined> {
+		const chunks = await this.chunkFile(fileContents) // todo handle empty files or very short files!
+		if (!chunks) {
+			console.error('failed to get chunks from file - file entry failing')
+			return
 		}
-
-		results.sort((a, b) => {
-			return a[0]> b[0] ? -1: 1
-		})
-
-		const result: Iterable<[string, number]> = results
-			.splice(0, resultNumber)
-			.filter(entry => entry[0] > relevanceThreshold)
-			.map((value) => {
-				return [value[1], value[0]]
+		const fileEmbeddingPromise = this.createEmbeddingBatch([fileContents])
+		const cEP: Promise<Chunk | undefined>[] = chunks.map(c => {
+			const p: Promise<Chunk | undefined> = this.createEmbeddingBatch([c]).then(e => {
+				if (e === undefined) return undefined
+				return ({
+					contents: c,
+					embedding: e.data[0].embedding
+				});
 			})
-
-		return new Map(result)
+			return p
+		})
+		const chunksWithEmbeddings: Array<Chunk | undefined> = await Promise.all(cEP)
+		const chunksNotUndefined: Chunk[] = chunksWithEmbeddings.filter(isChunk)
+		const fileEmbedding = await fileEmbeddingPromise
+		if (!fileEmbedding) {
+			console.error('failed to embed file')
+			return
+		}
+		return {
+			eTs: Date.now(),
+			mTs: mtime,
+			md5hash: md5hash,
+			fileEmbedding: fileEmbedding.data[0].embedding,
+			chunks: chunksNotUndefined
+		}
 	}
 
-	getByFilePath(filePath: string): Vector | undefined {
-		return this.filePathToVector.get(filePath)
+	async addFile(file: TFile) {
+		const existingFile = this.embeddings.get(file.path)
+		if (existingFile) return
+		const fileContents = await this.vault.read(file)
+		const hash = this.generateMd5Hash(fileContents)
+		const fileEntry = await this.generateFileEntry(fileContents, file.stat.mtime, hash)
+		if (fileEntry) {
+			this.embeddings.set(file.path, fileEntry)
+		}
+		await this.saveEmbeddingsToDatabaseFile() // todo debounce
 	}
 
-	async addVector(filePath: string, vector: Vector) {
-		this.filePathToVector.set(filePath, vector)
-		await this.saveVectorFile()
+	async updateFile(file: TFile) {
+		const existingFile = this.embeddings.get(file.path)
+		if (existingFile) {
+			const fileContents = await this.vault.read(file)
+			const hash = this.generateMd5Hash(fileContents)
+			if (hash !== existingFile.md5hash) {
+				const fileEntry = await this.generateFileEntry(fileContents, file.stat.mtime, hash)
+				if (fileEntry) {
+					this.embeddings.set(file.path, fileEntry)
+				}
+			} else {
+				console.log('no change, not updating db')
+			}
+		} else {
+			await this.addFile(file)
+		}
+		await this.saveEmbeddingsToDatabaseFile() // todo debounce
```

Comment:
We save in lots of places. I was thinking in general debouncing the save function would be good so we are only writing to the file periodically. Since we are working in memory its not imperative to have it save right away.